### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by markphelps** [flipt-io/flipt#760] [Ideas]
## What's Next?

I'd like to poll the community to see what they would like to see added to Flipt. I plan to create a public roadmap to better hold myself accountable and to enable focusing on features that the community wants/

Please upvote features you would like to see to help with prioritization. Im going to add a few that I think would be beneficial, but please feel free to add your own.

**Comment by markphelps** [comment]
### Built In Multiple Environment support

At least two environments out of the box:

- Development
- Production

Ideas:
- Everything would be duplicated at creation time across both environments
- You would have the ability to manage flags in each environment independantly of the other
- At evaluation time you would need to specify which environment you want to evaluate againts

**Comment by markphelps** [comment]
### Tagging

- The ability to add arbitrary tags to Flags/Segments, similar to how you can add tags to GH issues
- The ability to filter/search by tags

**Comment by markphelps** [comment]
### Basic Authz

- Either something like described in https://github.com/markphelps/flipt/discussions/733, where management (CRUD) runs on different port than evaluation
- OR basic API Key support where you can define which (if any) API Keys should be allowed to manage features
- OR perhaps both? Would love feedback

**Reply by tsickert** [reply]
I'd love to have support for JWT based auth.

I think the tricky part with JWT support is how the claim validation process would work. While verifying a signature is half the battle, the other half would be questions like:

- is the `sub` always the `entityId`?
- how do I tell flipt which claims/scopes I care about
- how do I assign those scopes by endpoint type (management vs client)? or each endpoint one by one? 
- how do I provide/manage the public key?

**Reply by markphelps** [reply]
Im wondering if API Token auth for API requests combined with OAuth for UI would be enough right now, similar to how GitHub PATs work. See #779

**Comment by markphelps** [comment]
### Support for Distributed Caches

- Redis
- Memached

Come to mind. This will allow Flipt caching to be used in an HA environment

**Reply by markphelps** [reply]
Redis support is done in #954 and released in v1.10

Will look to see if memcached support would be quick to add

**Comment by tsickert** [comment]
### OAuth Support In the Portal

I know this constitues a huge increase in scope for Flipt, but I'm interested in the feasability of adding OAuth support for logging into the Flipt portal. 

The OAuth flow is simple enough and the resulting token can be stored in local storage on the portal and then used when making API calls. The JWT feature from the discussion in the [Authz section](https://github.com/markphelps/flipt/discussions/760#discussioncomment-2398095) would then allow Flipt to verify and consume the claims from the token.

**Reply by markphelps** [reply]
This seems to me like it would provide the best value for Flipt users out of all the proposed features on the roadmap, so I think maybe we should just tackle it first, even if it is likely the most work out of all of the above.

I went ahead and created https://github.com/markphelps/flipt/issues/779 to track/document current thinking in order to get started. Would love feedback
